### PR TITLE
Allowed passing parameters into lazy initializer.

### PR DESCRIPTION
**Pull request by dkorolev**
Hi @mzhurovich,

Added a clean way to allow semi-lazy initialization.

PTAL once you're going through all the e-mail. :-)

Thanks!
Dima

### CODE REVIEW

File: Bricks/util/lazy_instantiation.h [hunk 1]
```diff
@@ -33,56 +33,118 @@ SOFTWARE.
...
+  // InstantiateAsSharedPtr as a `shared_ptr<T>`.
+  std::shared_ptr<T> InstantiateAsSharedPtr() const { return impl_->InstantiateAsSharedPtr(); }
+  template <typename TT = T_EXTRA_PARAMETER>
+  std::shared_ptr<T> InstantiateAsSharedPtrWithExtraParameter(const TT& parameter) const {
+    return impl_->InstantiateAsSharedPtr(parameter);
+  }
 
-  // Instantiate as a `shared_ptr<T>`.
-  std::shared_ptr<T> Instantiate() const { return impl_->Instantiate(); }
+  // InstantiateAsSharedPtr as a `unique_ptr<T>`.
```
Comment (mzhurovich) [inline Bricks/util/lazy_instantiation.h]: s/InstantiateAsSharedPtr/InstantiateAsUniquePtr/

File: Bricks/util/test.cc [hunk 2]
```diff
@@ -453,37 +454,87 @@ TEST(Util, LazyInstantiation) {
...
+  EXPECT_EQ(1, a_1.InstantiateAsSharedPtr()->foo);
+  EXPECT_EQ(2, a_2.InstantiateAsSharedPtr()->foo);
+  EXPECT_EQ(3, a_3.InstantiateAsSharedPtr()->foo);
+
+  EXPECT_EQ(1, b_1.InstantiateAsSharedPtr()->foo);
+  EXPECT_EQ(2, b_2.InstantiateAsSharedPtr()->foo);
+  EXPECT_EQ(3, b_3.InstantiateAsSharedPtr()->foo);
+
+  int q = 0;
+
```
Comment (mzhurovich) [inline Bricks/util/test.cc]: Remove linebreak?

File: Bricks/util/test.cc [hunk 3]
```diff
@@ -453,37 +454,87 @@ TEST(Util, LazyInstantiation) {
...
+  const auto bar_1_q = DelayedInstantiate<Bar>(1, std::ref(q));
+
+  q = 2;
+  EXPECT_EQ("1:2", bar_1_q.InstantiateAsSharedPtr()->AsString());
+  q = 3;
+  EXPECT_EQ("1:3", bar_1_q.InstantiateAsSharedPtr()->AsString());
+
+  const auto bar_x_q = DelayedInstantiateWithExtraParameter<Bar, int>(std::cref(q));
```
Comment (mzhurovich) [inline Bricks/util/test.cc]: Add linebreak? ;)

File: Bricks/util/test.cc [hunk 4]
```diff
@@ -453,37 +454,87 @@ TEST(Util, LazyInstantiation) {
...
+  EXPECT_EQ(3, a_3.InstantiateAsSharedPtr()->foo);
+
+  EXPECT_EQ(1, b_1.InstantiateAsSharedPtr()->foo);
+  EXPECT_EQ(2, b_2.InstantiateAsSharedPtr()->foo);
+  EXPECT_EQ(3, b_3.InstantiateAsSharedPtr()->foo);
+
+  int q = 0;
+
+  const auto bar_1_q = DelayedInstantiate<Bar>(1, std::ref(q));
+
```
Comment (mzhurovich) [inline Bricks/util/test.cc]: Or remove it here? :)

**Comment by mzhurovich** [comment]
@dkorolev Code LGTM! What is this special case with extra parameter for?

File: Bricks/util/lazy_instantiation.h (hunk 1 above)
Comment (dkorolev) [inline Bricks/util/lazy_instantiation.h]: :+1: renamed in 79c826f.

File: Bricks/util/test.cc (hunk 2 above)
Comment (dkorolev) [inline Bricks/util/test.cc]: 79c826f :)

File: Bricks/util/test.cc (hunk 3 above)
Comment (dkorolev) [inline Bricks/util/test.cc]: 79c826f :)

File: Bricks/util/test.cc (hunk 4 above)
Comment (dkorolev) [inline Bricks/util/test.cc]: Added the one above. :)

**Pull request merged**